feat: Add helper functions for permissions to std

kitsonk opened the pull request:
Provides two helper functions which make it simpler to deal with permissions.  Makes it easier to have single statements dealing with permissions in the code.

`grant` will resolve with `true` if the permission is granted, prompting if needed, and resolve with `false` if the permission is denied:

```ts
if (await grant({ name: "net" }) {
  // ... some cool networking thing...
} else {
  console.error("Really need access to the net to do this cool thing");
  Deno.exit(1);
}
```

`grantOrThrow` will simply throw if the permission can't be granted, otherwise just resolve:

```ts
await grantOrThrow({ name: "net" });
// ... some cool networking thing ...
```

kitsonk commented:
I don't know if there is any way currently to really test with variations in permissions, which limits the testing that can be done on the code.

bartlomieju commented:
> I don't know if there is any way currently to really test with variations in permissions, which limits the testing that can be done on the code.

At the moment there's is no way to use different permissions for std, but that should be possible in the future (working on first class support for `{ perms: {} }` option for tests

nayeemrmn commented:
This should take a list of permission descriptors, that's what would actually provide value.

kitsonk commented:
@nayeemrmn I've updated it, `grant()` and `grantOrThrow()` now accepts a spread of descriptors or an array of descriptors.  I do think that the single perm is still useful, so I added `isGranted()` which can be used.

I just realised though, that if no permissions are granted, I shouldn't return a length array.  Actually `isGranted` is redundant if I resolve `void` if no permissions are granted...  Going to update again.

kitsonk commented:
So yeah, for single perms, you can do this:

```ts
if (await grant({ name: "net" }) {
 // per is granted
}
```

But for multiple perms you would want to do something else, like:

```ts
const perms = await grant({ name: "net" }, { name: "read" });
if (perms.length === 2) {
  // all perms granted
}
```

nayeemrmn commented:
I'm not sure about this... when this was requested the idea was to have a utility that would turn this:
```js
const errors = [];
if ((await Deno.permissions.request({ name: "net" })).state != "granted") {
  errors.push("- net");
}
if ((await Deno.permissions.request({ name: "read", path: "./subdir1" })).state != "granted") {
  errors.push("- read: ./subdir1");
}
if ((await Deno.permissions.request({ name: "write", path: "./subdir2" })).state != "granted") {
  errors.push("- write: ./subdir2");
}
if (errors.length > 0 ) {
  throw new Deno.errors.PermissionDenied(`The following permissions are required but not granted:\n${errors.join("\n")}`);
}
// Do something cool with net, read at ./subdir1, write at subdir2.
```
into this:
```js
requestPermissions([
  { name: "net" },
  { name: "read", path: "./subdir1" },
  { name: "write", path: "./subdir2" }
]);
// Do something cool with net, read at ./subdir1, write at subdir2.
```
To list all permissions that are required and missing, not just the first to be discovered. The utility should address this use case and as little else as possible. We just seem to be getting rid of some boilerplate with this.

kitsonk commented:
> The utility should address this use case and as little else as possible.

In your opinion.  I am not writing code to meet your use case specifically.  I am writing code that meets a use case I have run into.  Just to be clear, I am not taking an order here.  My whole purpose was to get rid of boilerplate.  As written `grant()` effectively does that.  It makes sense to change `grantOrThrow()` to throw mentioning all missing.

nayeemrmn commented:
My mistake, I assumed this was in response to discussion starting here https://github.com/denoland/deno/issues/3811#issuecomment-579304951 but I guess it just came up for you again. `grantOrThrow()` pretty much is now but I don't see why it's named so secondarily when it's far less niche than the other one, that seemed odd.

How about `grantOrThrow()` -> `grant()`, `grant()` -> `maybeGrant()`?

kitsonk commented:
Personally I would want to get wider feedback, as I really don't like throwing, as uncaught errors are usually not very user friendly, and `try ... catch` doesn't really provide any value.  I much rather have a branch where I would `if (!grant(...)) { /* some graceful handling and exit */ }` or log a message and degrade the experience if I can then to just log out a stack trace which is meaningless to the user.

hazae41 commented:
How about 

    grant(name: string, path?: string)
    grant("net")
    grant("read", "./plugins")
    grant("write", "./export")

or 

    grant(...perms: [string, string])
    grant(["net"], ["read", "./plugins"], ["write", "./export"])

kitsonk commented:
@hazae41 I would rather keep the descriptor, as it future proofs any other permission changes in the future.  It might not always be just a name with an optional second string.  Objects are also much more clearer than tuples.

nayeemrmn commented on std/permissions/mod.ts:
```diff
@@ -0,0 +1,115 @@
+    let state = (await Deno.permissions.query(descriptor)).state;
+    if (state === "prompt") {
+      state = (await Deno.permissions.request(descriptor)).state;
+    }
```
Is there a difference between this and just `const state = (await Deno.permissions.request(descriptor)).state;`? I believe `request()` handles every situation intuitively.

kitsonk replied:
Good point...

nayeemrmn commented on std/permissions/mod.ts:
```diff
@@ -0,0 +1,115 @@
+        case "net":
+          return pd.url ? `${pd.name}(${pd.url})` : pd.url;
+        default:
+          return pd.name;
```
It could be really helpful for some people to stringify them to the CLI flags to get an error like
```
The following permissions have not been granted:
  --allow-env
  --allow-read=path/to/dir
```

kitsonk replied:
Yeah, that is better output...

ry commented:
@bartlomieju Looks good - feel free to land after @nayeemrmn's concerns are discussed.

@kitsonk sorry about the delay on this one!

kitsonk commented:
Ok, PR updated... The output is much better when it throws.  Thanks for the suggestion @nayeemrmn.

bartlomieju commented:
@kitsonk I'll land this PR after we fix #4373
